src/main/scala/com/twitter/finagle/websocket/WebSocketHandler.scala

```diff
@@ -113,91 +50,50 @@ private[finagle] class WebSocketServerHandler extends SimpleChannelUpstreamHandl
+private[finagle] class WebSocketClientHandler extends SimpleChannelHandler {
+  @volatile private[this] var ref: Option[WebSocketClientHandshaker] = None
 
-  override def messageReceived(ctx: ChannelHandlerContext, e: MessageEvent) = {
+  override def messageReceived(ctx: ChannelHandlerContext, e: MessageEvent): Unit =
     e.getMessage match {
-      case res: HttpResponse if handshaker.isDefined =>
-        val hs = handshaker.get
-        if (!hs.isHandshakeComplete)
-          hs.finishHandshake(ctx.getChannel, res)
-
-      case frame: CloseWebSocketFrame =>
-        ctx.getChannel.close()
-
-      case frame: PingWebSocketFrame =>
-        ctx.getChannel.write(new PongWebSocketFrame(frame.getBinaryData))
-
-      case frame: PongWebSocketFrame =>
-        // do nothing
-
-      case frame: TextWebSocketFrame =>
-        val ch = ctx.getChannel
-        ch.setReadable(false)
-        (messagesBroker ! frame.getText) ensure { ch.setReadable(true) }
+      case res: HttpResponse =>
+        ref match {
+          case None =>
+            throw new IllegalStateException("unexpected HTTP response before handshake")
+          case Some(handshaker) if handshaker.isHandshakeComplete =>
+            throw new IllegalStateException("unexpected HTTP response after handshake")
+          case Some(handshaker) =>
+            handshaker.finishHandshake(ctx.getChannel, res)
+        }
 
-      case frame: BinaryWebSocketFrame =>
-        val ch = ctx.getChannel
-        ch.setReadable(false)
-        (binaryMessagesBroker ! frame.getBinaryData.array) ensure { ch.setReadable(true) }
+      case frame: WebSocketFrame =>
+        Channels.fireMessageReceived(ctx, frame)
 
       case invalid =>
         Channels.fireExceptionCaught(ctx,
-          new IllegalArgumentException("invalid message \"%s\"".format(invalid)))
+          new IllegalArgumentException(s"invalid message: $invalid"))
     }
-  }
-
 
-  override def writeRequested(ctx: ChannelHandlerContext, e: MessageEvent) = {
+  override def writeRequested(ctx: ChannelHandlerContext, e: MessageEvent): Unit =
     e.getMessage match {
-      case sock: WebSocket =>
-        write(ctx, sock)
-
-        def close() {
-          keepAliveTask.foreach(_.close())
-          Channels.close(ctx.getChannel)
-        }
-
-        val webSocket = sock.copy(
-          messages = messagesBroker.recv,
-          binaryMessages = binaryMessagesBroker.recv,
-          onClose = closer,
-          close = close)
-
-        Channels.fireMessageReceived(ctx, webSocket)
-
-        val wsFactory = new WebSocketClientHandshakerFactory
-        val hs = wsFactory.newHandshaker(sock.uri, sock.version, null, false, sock.headers)
-        handshaker = Some(hs)
-        hs.handshake(ctx.getChannel).addListener(new ChannelFutureListener {
-          override def operationComplete(f:ChannelFuture) {
-
-            // initiate the keep-alive
-            for(interval <- sock.keepAlive) {
```

Comment:
My thinking here was that since we are exposing the Ping/Pong frames, users can do this themselves.